transaction/core/tests/input_rules.rs

```diff
@@ -63,25 +72,290 @@ fn test_input_rules_verify_max_tombstone() {
 
     // Modify the Tx to have some (empty) input rules.
     // (This invalidates the signature, but we aren't checking that here)
-    tx.prefix.inputs[0].input_rules = Some(InputRules {
-        required_outputs: vec![],
-        max_tombstone_block: 0,
-    });
+    tx.prefix.inputs[0].input_rules = Some(InputRules::default());
 
     // Check that the Tx is following input rules (vacuously)
     get_first_rules(&tx).verify(block_version, &tx).unwrap();
 
     // Declare the tombstone block limit to be one less than the current value.
-    tx.prefix.inputs[0].input_rules = Some(InputRules {
-        required_outputs: vec![],
-        max_tombstone_block: tx.prefix.tombstone_block - 1,
-    });
+    get_first_rules_mut(&mut tx).max_tombstone_block = tx.prefix.tombstone_block - 1;
 
-    assert!(get_first_rules(&tx).verify(block_version, &tx).is_err());
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::MaxTombstoneBlockExceeded)
+    );
 
     // Set the tombstone block limit to be more permissive, now everything should be
     // good
     get_first_rules_mut(&mut tx).max_tombstone_block = tx.prefix.tombstone_block + 1;
 
     get_first_rules(&tx).verify(block_version, &tx).unwrap();
 }
+
+fn change_committed_amount(r_txo: &RevealedTxOut, new_amount: Amount) -> RevealedTxOut {
+    // Confirm that this can even be revealed
+    r_txo.reveal_amount().unwrap();
+
+    let mut result = r_txo.clone();
+    let new_masked_amount = MaskedAmountV2::new_from_amount_shared_secret(
+        new_amount,
+        &r_txo.amount_shared_secret[..].try_into().unwrap(),
+    )
+    .unwrap();
+
+    // Confirm that the new masked amount can be decoded using this shared secret as
+    // expected
+    assert_eq!(
+        new_amount,
+        new_masked_amount
+            .get_value_from_amount_shared_secret(
+                &r_txo.amount_shared_secret[..].try_into().unwrap()
+            )
+            .unwrap()
+            .0
+    );
+
+    result.tx_out.masked_amount = Some(MaskedAmount::V2(new_masked_amount));
+    result
+}
+
+// Test that input rules verification is working for fractional output rules
+#[test]
+fn test_input_rules_verify_fractional_outputs() {
```

Comment:
> individual unit tests for each case and or suggested function

I don't think it is a good idea to test the suggested functions instead of `verify`. I want to ensure very strong test coverage on the `verify` API because that is the actual public API that is called by transaction validation. Testing only its internals leaves open the possibility that a logic mistake in the overall function (perhaps a call to one of the internals is skipped?) leads to a bug in transaction validation.

I'm okay with the idea of splitting the recipe test into a few smaller tests. However, I've been getting a lot of feedback that there are too many LOC in the PR. Splitting things up into more tests, and repeating all the test boilerplate, will increase the LOC of the PR. So I'm reluctant to do too much of that.

Reply:
> Testing only its internals leaves open the possibility that a logic mistake in the overall function
This is a very real concern.  I think ideally one would have _at least_ 4 tests of `verify()` that exercised:
1. Happy path
2. Failure on tombstone
2. Failure on required outputs
3. Failure on partial fill rules
One could add a couple more per failure as an extra feel good.
Then there would be full logic coverage with the unit tests of each sub function.

Often it feels like it's unit vs integration test.  I don't see it as one or the other. They both should exist and they help to support each other.  Integration tests support higher level ideas, while not needing to robustly exercise all the edge cases.  Unit tests get into the nitty gritty, helping prevent off by one errors, correct loop terminations, proper unit conversions, etc.
It can feel like wasted effort because the higher level integration tests seem to be testing the same or similar things as the unit tests, but like the `verify()` breakup suggestion, the integration test is ensuring that `verify()` delegates to the proper functionality and that those pieces of functionality work together, while the unit tests are robustly ensuring the functionality.